### ISSUE TITLE
Make rest.vertx more extensible

### ISSUE THREAD

**Issue by gofabian** [zandero/rest.vertx#41]
At first thank you for your project. I really like it because it is a good alternative to the official Vert.x Web Contract library and enables me to start a project rapidly.

I use rest.vertx with Kotlin and would like to have support for Kotlin coroutines. At the moment it is not possible to change some behaviours of rest.vertx without forking your project. I suggest to change the structure a bit so that support for Kotlin coroutines or Quasar fibers will be possible as an optional external library.

I would like to support something like this (this will not work at the moment):

<pre>
@POST
@Path("/echo")
<b>suspend</b> fun echoFunction(echo: Echo): Echo {
    // ... do something in an asynchronous way but in synchronous coroutine style ...
    return echo
}
</pre>

But at the moment I can only do this (with some code around the coroutine):

<pre>
@POST
@Path("/echo")
fun echoFunction(echo: Echo): <b>Future&lt;Echo&gt;</b> {
    <b>Future<Echo> future = Future.future()

    // launch coroutine
    GlobalScope.launch(vertx.dispatcher)</b> {
        // ... do something in an asynchronous way but in synchronous coroutine style ...
        future.complete(echo)
    <b>}

    return future</b>
}
</pre>

The first example is much more concise. In order to enable it from external code I need to open rest.vertx a bit:

- The AnnotationProcessor methods would be changed from static to instance methods. A few methods would be changed from private to protected.
- A RouteHandlerFactory would be introduced that creates Handler<RoutingContext> for each operation.

Both would be members of RestRouter. The RestRouter would have a possibilty to overwrite the AnnotationProcessor by something like a sub class and a new RouterHandlerFactory can be added that would for example support Kotlin coroutines or Quasar fibers.

What do you think about such a change? I think rest.vertx would be even better if we made it a bit more extensible.

**Comment by drejc** [comment]
It seems to me that you already have figured out the solution, so just go
ahead, fork and create a patch.

I will than merge and release a new version.

**Comment by YounesRahimi** [comment]
Because of a need for adding kotlin dependencies, all these changes need to happen in an extra library which can use Rest.VertX. I have created and published a library with such requirements named [rest.vertx-kotlin-coroutines](https://github.com/YounesRahimi/rest.vertx-kotlin-coroutines) which have same usage API for kotlin but with a `suspend` modifier. The full documentation is available at the project Github page.

**Comment by YounesRahimi** [comment]
As @gofabian mentioned there are some needs for changing in rest.vertx SPI to support extensibility of the library. For project [rest.vertx-kotlin-coroutines](https://github.com/YounesRahimi/rest.vertx-kotlin-coroutines) I forced to create an extra RestBuilder and RestRouter which have the same API with rest.vertx but can register and build suspendable routes. I also forced to access the private fields of RouteDefinition using reflection which is a bad practice which can be avoided by making some key fields and methods protected.
